Game/Source/GDKTestGyms/Private/UserExperienceComponent.cpp

```diff
@@ -0,0 +1,184 @@
+UUserExperienceComponent::UUserExperienceComponent()
+{
+	// Set this component to be initialized when the game starts, and to be ticked every frame.  You can turn these features
+	// off to improve performance if you don't need them.
+	PrimaryComponentTick.bCanEverTick = true;
+	bWantsInitializeComponent = true;
+	bAutoActivate = true;
+}
+
+void UUserExperienceComponent::InitializeComponent()
+{
+	ElapsedTime = 0.0f;
+	ServerClientRTT = 0.0f;
+	ServerViewLateness = 0.0f;
+	ClientRTTTimer = 0.0f; 
+	SetIsReplicated(true);
+	UActorComponent::InitializeComponent();
+}
+
+void UUserExperienceComponent::SendServerRPC()
+{
+	ServerRTT(ElapsedTime);
+}
+
+void UUserExperienceComponent::UpdateClientObservations(float DeltaTime)
+{
+	// Prune out any components which went out of view to avoid skewing results
+	TMap<UUserExperienceComponent*, ObservedUpdate> NewObservations;
+	for (TObjectIterator<UUserExperienceComponent> It; It; ++It)
+	{
+		if (this != *It && It->GetOwner() && It->GetOwner()->GetWorld() == GetWorld())
+		{
+			if (ObservedUpdate* PreviousUpdate = ObservedComponents.Find(*It))
+			{
+				// Carry over into new observations
+				NewObservations.Add(*It, MoveTemp(*PreviousUpdate));
+			}
+			else
+			{
+				NewObservations.Add(*It, ObservedUpdate{});
+			}
+		}
+	}
+	ObservedComponents = NewObservations;
+
+	for (TPair<UUserExperienceComponent*, ObservedUpdate>& ObservationPair : ObservedComponents)
+	{
+		ObservedUpdate& Observation = ObservationPair.Value;
+		UUserExperienceComponent* Component = ObservationPair.Key;
+
+		Observation.TimeSinceChange += DeltaTime;
+		if (Observation.Value != Component->ClientTime)
+		{
+			// New value observed.
+			Observation.Value = Component->ClientTime;
+			Observation.TrackedChanges.Push(Observation.TimeSinceChange);
+			Observation.TimeSinceChange = 0.0f;
+
+			if (Observation.TrackedChanges.Num() > NumWindowSamples)
+			{
+				Observation.TrackedChanges.RemoveAt(0);
+			}
+		}
+	}
+}
+
+float UUserExperienceComponent::CalculateWorldFrequency()
+{
+	float Frequency = 0.0f;
+	for (TPair<UUserExperienceComponent*, ObservedUpdate>& ObservationPair : ObservedComponents)
+	{
+		ObservedUpdate& Observation = ObservationPair.Value;
+		UUserExperienceComponent* Component = ObservationPair.Key;
+
+		if (!Component->IsActive())
+		{
+			continue;
+		}
+
+		if (Observation.TrackedChanges.Num() == NumWindowSamples)
+		{
+			float AverageUpdateRate = Calculate80thPctAverage(Observation.TrackedChanges);
+			Frequency += AverageUpdateRate*1000.0f;
+		}
+		else
+		{
+			Frequency += 0.0f;
+		}
+	}
+	
+	Frequency /= static_cast<float>(ObservedComponents.Num());
+	return Frequency;
+}
+
+// Called every frame
+void UUserExperienceComponent::TickComponent(float DeltaTime, ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction)
+{
+	Super::TickComponent(DeltaTime, TickType, ThisTickFunction);
+	
+	if (GetOwner()->GetLocalRole() == ROLE_AutonomousProxy)
+	{
+		ClientRTTTimer -= DeltaTime;
```

Comment:
I think I'd prefer a system where the client's initial execution is tied to the completion of the last trace. That way we can potentially fail if the rpcs start getting queued for instance.

Reply:
This now happens with StartRoundtrip and EndRoundtrip (the data is stored in a map). It works similar to before though, if it takes longer than a second to come back then EndRoundtrip is triggered by EndRoundtrip (the previous key still being active)